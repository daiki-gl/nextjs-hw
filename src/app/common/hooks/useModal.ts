import { useState } from 'react'

/*
  モーダルの開閉用
  @returns isOpen 開いているかのチェック
  @returns setIsOpen isOpenのtrue/false更新用関数
  @returns closeModal モーダルを閉じる関数
*/
export default function useModal() {
  const [isOpen, setIsOpen] = useState(false)

  const openModal = () => {
    setIsOpen(true)
  }
  const closeModal = () => setIsOpen(false)

  return { isOpen, openModal, closeModal }
}
