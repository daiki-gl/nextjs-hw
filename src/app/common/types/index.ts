export interface UserData {
  id: number
  name: string
  phone: string
  address: string
  status: boolean
  payment: string
}

export interface ButtonProps {
  text: string
  onClick?: React.MouseEventHandler<HTMLButtonElement>
  disabled?: boolean
  outline?: boolean
  noDeco?: boolean
  type?: 'button' | 'submit' | 'reset'
}
export interface ModalProps {
  isOpen: boolean
  onClose: () => void
  children: React.ReactNode
}

export interface useSelectType {
  selectedAll: boolean | 'indeterminate'
  setSelectedAll: (value: boolean | 'indeterminate') => void
  setSelected: (value: boolean[]) => void
  selected: boolean[]
}

export interface ShowDataContextType {
  showAddUserData: UserData[]
  setShowAddUserData: React.Dispatch<React.SetStateAction<UserData[]>>
}

export interface TypeContextType {
  type: string | null
  setType: React.Dispatch<React.SetStateAction<string | null>>
}

export interface SearchResultContextType {
  searchResult: UserData[] | null
  setSearchResult: React.Dispatch<React.SetStateAction<UserData[]>>
}
