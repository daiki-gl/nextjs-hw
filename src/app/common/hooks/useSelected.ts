import { SetStateAction, useEffect, useState } from 'react'
import { UserData } from '../types'

/*
  ユーザーをidから検索
  @param setUserDetails ユーザーの詳細情報をsetする関数
  
  ユーザーのチェックボックス用
  @returns selected ユーザー毎のチェックボックスのcheckedの確認
  @returns selectedAll showUserDataのユーザー全体のcheckedの確認
  @returns setSelected selectedの更新用
  @returns setSelectedAll selectedAllの更新用
  */
export default function useSelected(showUserData: UserData[] | null) {
  const [selected, setSelected] = useState<boolean[]>([])
  const [selectedAll, setSelectedAll] = useState<boolean | 'indeterminate'>(
    false
  )

  /*
  showUserDataの更新時にselectedのcheckの有無の確認と更新
  */
  useEffect(() => {
    if (showUserData && showUserData.length > 0) {
      const newSelected = showUserData.map(() => false) //shoUserDataにfalseを割り当てている
      setSelected(newSelected)
    } else {
      setSelected([])
    }
  }, [showUserData])

  /*
  selectedAllをtrue/falseに更新
  */
  useEffect(() => {
    const selectedResult = selected.filter((select: boolean) => select == true) // trueのユーザーをfilter
    if (
      showUserData &&
      showUserData.length > 0 &&
      selectedResult.length == showUserData?.length
    ) {
      setSelectedAll(true)
    } else {
      setSelectedAll(false)
    }
  }, [selected, showUserData])

  /*
  ユーザーをidから検索
  @param id ユーザーID
  @param setUserDetails ユーザーの詳細情報をsetする関数
  */
  function fetchUserData(
    id: number,
    setUserDetails: React.Dispatch<SetStateAction<UserData | undefined>>
  ): void {
    const user = showUserData?.find((user) => user.id === id)
    setUserDetails(user)
  }

  return { selected, setSelected, selectedAll, setSelectedAll, fetchUserData }
}
