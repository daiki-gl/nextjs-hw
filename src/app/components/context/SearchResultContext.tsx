'use client';
import { SearchResultContextType, UserData } from '@/app/common/types';
import { createContext, useContext, useState } from 'react';



const SearchResultContext = createContext<SearchResultContextType | null>(null);

export const SearchResultProvider = ({ children }: { children: React.ReactNode }) => {
  const [searchResult, setSearchResult] = useState<UserData[]>([]);

  return (
    <SearchResultContext.Provider value={{ searchResult, setSearchResult }}>
      {children}
    </SearchResultContext.Provider>
  );
};

export const useSearchResultContext = () => {
  const context = useContext(SearchResultContext);
  if (!context) throw new Error('useShowDataContext must be used within ShowDataProvider');
  return context;
};