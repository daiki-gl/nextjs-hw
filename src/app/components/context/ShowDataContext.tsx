'use client';
import { createContext, useContext, useState } from 'react';
import { ShowDataContextType, UserData } from '../../common/types';


const ShowDataContext = createContext<ShowDataContextType | null>(null);

export const ShowDataProvider = ({ children }: { children: React.ReactNode }) => {
  const [showAddUserData, setShowAddUserData] = useState<UserData[]>([]);

  return (
    <ShowDataContext.Provider value={{ showAddUserData, setShowAddUserData }}>
      {children}
    </ShowDataContext.Provider>
  );
};

export const useShowDataContext = () => {
  const context = useContext(ShowDataContext);
  if (!context) throw new Error(' must be used within ShowDataProvider');
  return context;
};