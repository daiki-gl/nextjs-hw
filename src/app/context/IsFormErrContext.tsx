'use client';
import { createContext, useContext, useState } from 'react';
import { IsFormErrContextType } from '../common/types';


const IsFormErrContext = createContext<IsFormErrContextType | null>(null);

export const IsFormErrContextProvider = ({ children }: { children: React.ReactNode }) => {
  const [isFormErr, setIsFormErr] = useState<boolean>(true);

  return (
    <IsFormErrContext.Provider value={{ isFormErr, setIsFormErr }}>
      {children}
    </IsFormErrContext.Provider>
  );
};

export const useIsFormErrContext = () => {
  const context = useContext(IsFormErrContext);
  if (!context) throw new Error('useShowDataContext must be used within ShowDataProvider');
  return context;
};