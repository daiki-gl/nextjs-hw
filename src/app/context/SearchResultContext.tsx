'use client';
import { createContext, useContext, useState } from 'react';
import { SearchResultContextType, UserData } from '../common/types';


const SearchResultContext = createContext<SearchResultContextType | null>(null);

export const SearchResultProvider = ({ children }: { children: React.ReactNode }) => {
  const [searchResult, setSearchResult] = useState<UserData[]>([]);

  return (
    <SearchResultContext.Provider value={{ searchResult, setSearchResult }}>
      {children}
    </SearchResultContext.Provider>
  );
};

export const useSearchResultContext = () => {
  const context = useContext(SearchResultContext);
  if (!context) throw new Error('useShowDataContext must be used within ShowDataProvider');
  return context;
};