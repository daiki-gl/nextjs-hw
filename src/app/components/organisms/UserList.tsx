'use client';
import { SetStateAction } from "react";
import { UserData } from "../../common/types";
import useSelected from "@/app/common/hooks/useSelected";
import TableHeader from "../molecules/TableHeader";
import TableBody from "../molecules/TableBody";
import Button from "../atoms/Button";


export default function UserList({showUserData, setShowUserData}
    :{
        showUserData: UserData[] | null; 
        setShowUserData: React.Dispatch<SetStateAction<UserData[]>>;
    } ) {
   const {selected,selectedAll,setSelected,setSelectedAll} = useSelected(showUserData);
        
    return (
        <div className="w-3/5 mx-auto my-10">
            <h1 className="text-2xl font-bold">検索結果 *{showUserData && showUserData.length}件</h1>
            <table className="w-full mt-5">
                <TableHeader select={{selectedAll,setSelectedAll,selected,setSelected}} />
                <TableBody showUserData={showUserData} selected={selected} setSelected={setSelected} />
            </table>
            
            <div className="w-full text-center my-7">
                    <Button text="閉じる" onClick={() => setShowUserData([])} />
                    <Button disabled={showUserData == null || showUserData.length == 0} text="追加" />
                </div>
        </div>
    )
}