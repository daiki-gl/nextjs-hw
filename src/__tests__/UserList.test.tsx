import { render, screen, waitFor, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import UserList from '../app/components/organisms/UserList';
import { useShowDataContext } from '../app/components/context/ShowDataContext';
import { useSearchResultContext } from '../app/components/context/SearchResultContext';
import { useIsFormErrContext } from '../app/components/context/IsFormErrContext';
import useSelected from '../app/common/hooks/useSelected';
import useModal from '../app/common/hooks/useModal';
import { useRouter } from 'next/navigation';
import { UserData } from '@/app/common/types';

// Mock data
const mockUsers = [
  { id: 1, name: '山田太郎', phone: '09012345678', address: '東京都', status: false, payment: '1000' },
  { id: 2, name: '佐藤花子', phone: '08098765432', address: '大阪府', status: false, payment: '2000' },
  { id: 3, name: '田中一郎', phone: '07011112222', address: '神奈川県', status: true, payment: '3000' },
];


jest.mock('../app/components/context/ShowDataContext', () => ({
  useShowDataContext: jest.fn(),
}));

jest.mock('../app/components/context/SearchResultContext', () => ({
  useSearchResultContext: jest.fn(),
}));

jest.mock('../app/components/context/IsFormErrContext', () => ({
  useIsFormErrContext: jest.fn(),
}));

jest.mock('../app/common/hooks/useSelected', () => ({
  __esModule: true,
  default: jest.fn(() => ({
    selected: {},
    toggleUserSelection: jest.fn(),
    toggleSelectAllCurrentPage: jest.fn(),
  })),
}));

jest.mock('../app/common/hooks/useModal', () => ({
  __esModule: true,
  default: jest.fn(() => ({
    isOpen: false,
    openModal: jest.fn(),
    closeModal: jest.fn(),
  })),
}));
jest.mock('next/navigation', () => ({
  useRouter: jest.fn(),
}));

jest.mock('../app/components/molecules/Pagenation', () => ({
  __esModule: true,
  default: ({ limit, showUserData }: { limit: number; showUserData: UserData[] }) => (
    <div data-testid="pagination-mock">
      {`Pagination: limit=${limit} data-length=${showUserData?.length}`}
    </div>
  ),
}));

describe('UserListコンポーネント', () => {
  const mockSetShowUserData = jest.fn();
  const mockRouterPush = jest.fn();
  const mockSetShowAddUserData = jest.fn();
  const mockToggleUserSelection = jest.fn();
  const mockToggleSelectAll = jest.fn();
  const mockOpenModal = jest.fn();
  const mockCloseModal = jest.fn();
  
  beforeEach(() => {
    jest.clearAllMocks();
    (useShowDataContext as jest.Mock).mockReturnValue({
      setShowAddUserData: mockSetShowAddUserData,
    });
    (useSearchResultContext as jest.Mock).mockReturnValue({
      searchResult: mockUsers,
    });
    (useIsFormErrContext as jest.Mock).mockReturnValue({
      isFormErr: false,
    });
    (useSelected as jest.Mock).mockReturnValue({
      selected: { 1: false, 2: false },
      toggleUserSelection: mockToggleUserSelection,
      toggleSelectAllCurrentPage: mockToggleSelectAll,
    });
    (useModal as jest.Mock).mockReturnValue({
      isOpen: false,
      openModal: mockOpenModal,
      closeModal: mockCloseModal,
    });
    (useRouter as jest.Mock).mockReturnValue({
      push: mockRouterPush,
    });
  });

  it('正しい検索結果件数とユーザーテーブルが表示されること', () => {
    render(<UserList showUserData={mockUsers} setShowUserData={mockSetShowUserData} />);
    expect(screen.getByText('検索結果 *3件')).toBeInTheDocument();
    expect(screen.getByRole('table')).toBeInTheDocument();
    expect(screen.getAllByRole('row').length).toBe(4); // 1 header + 3 users
  });

  it('データがないときに「表示するデータがありません」と表示されること', () => {
    render(<UserList showUserData={[]} setShowUserData={mockSetShowUserData} />);
    expect(screen.getByText('表示するデータがありません。')).toBeInTheDocument();
  });

  it('閉じるボタンをクリックするとsetShowUserDataが空の配列で呼び出されること', () => {
    render(<UserList showUserData={mockUsers} setShowUserData={mockSetShowUserData} />);
    fireEvent.click(screen.getByRole('button', { name: '閉じる' }));
    expect(mockSetShowUserData).toHaveBeenCalledWith([]);
  });

  it('ユーザーが選択されていない場合、追加ボタンが無効化されること', () => {
    render(<UserList showUserData={mockUsers} setShowUserData={mockSetShowUserData} />);
    const addButton = screen.getByRole('button', { name: '追加' });
    expect(addButton).toBeDisabled();
  });

  it('ユーザーが選択された場合、追加ボタンが有効化されること', () => {
    (useSelected as jest.Mock).mockReturnValue({
      selected: { 1: true, 2: false },
      toggleUserSelection: mockToggleUserSelection,
      toggleSelectAllCurrentPage: mockToggleSelectAll,
    });
    render(<UserList showUserData={mockUsers} setShowUserData={mockSetShowUserData} />);
    const addButton = screen.getByRole('button', { name: '追加' });
    expect(addButton).toBeEnabled();
  });

  it('追加ボタンをクリックすると/addページに遷移すること', () => {
    (useSelected as jest.Mock).mockReturnValue({
      selected: { 1: true, 2: false },
      toggleUserSelection: mockToggleUserSelection,
      toggleSelectAllCurrentPage: mockToggleSelectAll,
    });
    render(<UserList showUserData={mockUsers} setShowUserData={mockSetShowUserData} />);
    const addButton = screen.getByRole('button', { name: '追加' });
    fireEvent.click(addButton);
    expect(mockRouterPush).toHaveBeenCalledWith('/add');
  });

  it('名前をクリックするとモーダルが開き、ユーザー詳細がセットされること', async () => {
    (useModal as jest.Mock).mockReturnValue({
      isOpen: true,
      openModal: mockOpenModal,
      closeModal: mockCloseModal,
    });
    render(<UserList showUserData={mockUsers} setShowUserData={mockSetShowUserData} />);
    
    const firstUserName = screen.getByRole('button', { name: '山田太郎' });
    fireEvent.click(firstUserName);

    expect(mockOpenModal).toHaveBeenCalledTimes(1);

    // モーダルが表示されていることを確認
    await waitFor(() => {
      expect(screen.getByText('詳細情報')).toBeInTheDocument();
      expect(screen.getByText('名前：山田太郎')).toBeInTheDocument();
    });
  });

  it('ユーザーのチェックボックスをクリックするとtoggleUserSelectionフックが呼び出されること', () => {
    render(<UserList showUserData={mockUsers} setShowUserData={mockSetShowUserData} />);
    const checkboxes = screen.getAllByRole('checkbox');
    fireEvent.click(checkboxes[1]);
    expect(mockToggleUserSelection).toHaveBeenCalledWith(1, true);
  });

  it('スナップショットが一致すること', () => {
    const { asFragment } = render(<UserList showUserData={mockUsers} setShowUserData={mockSetShowUserData} />);
    expect(asFragment()).toMatchSnapshot();
  });
});