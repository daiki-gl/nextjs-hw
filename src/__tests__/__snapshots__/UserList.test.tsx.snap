// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`UserListコンポーネント スナップショットが一致すること 1`] = `
<DocumentFragment>
  <div
    class="w-3/5 mx-auto mt-10"
  >
    <h1
      class="text-2xl font-bold"
    >
      検索結果 *3件
    </h1>
    <table
      class="w-full mt-5 border-collapse border border-gray-300"
    >
      <thead>
        <tr
          class="border-b-2 border-gray-300"
        >
          <th
            class="py-3 px-4 text-center w-[70px]"
          >
            <input
              type="checkbox"
            />
          </th>
          <th
            class="text-center py-2 px-4 border border-gray-300 w-[180px]"
          >
            名前
          </th>
          <th
            class="text-center py-2 px-4 border border-gray-300 w-[213px]"
          >
            電話番号
          </th>
          <th
            class="text-center py-2 px-4 border border-gray-300 w-[150px]"
          >
            住所
          </th>
          <th
            class="text-center py-2 px-4 border border-gray-300 w-[170px]"
          >
            ステータス
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          class="border-b border-gray-300"
        >
          <td
            class="text-center py-2 px-4 border border-gray-300 w-[70px]"
          >
            <input
              type="checkbox"
            />
          </td>
          <td
            class="text-center py-2 px-4 border border-gray-300  w-[180px] hover:text-gray-400 underline underline-offset-2"
          >
            <button
              class="cursor-pointer w-full text-center"
            >
              山田太郎
            </button>
          </td>
          <td
            class="text-center py-2 px-4 border border-gray-300 w-[213px]"
          >
            09012345678
          </td>
          <td
            class="text-center py-2 px-4 border border-gray-300 w-[150px]"
          >
            東京都
          </td>
          <td
            class="text-center py-2 px-4 border border-gray-300 w-[170px]"
          >
            未
          </td>
        </tr>
        <tr
          class="border-b border-gray-300"
        >
          <td
            class="text-center py-2 px-4 border border-gray-300 w-[70px]"
          >
            <input
              type="checkbox"
            />
          </td>
          <td
            class="text-center py-2 px-4 border border-gray-300  w-[180px] hover:text-gray-400 underline underline-offset-2"
          >
            <button
              class="cursor-pointer w-full text-center"
            >
              佐藤花子
            </button>
          </td>
          <td
            class="text-center py-2 px-4 border border-gray-300 w-[213px]"
          >
            08098765432
          </td>
          <td
            class="text-center py-2 px-4 border border-gray-300 w-[150px]"
          >
            大阪府
          </td>
          <td
            class="text-center py-2 px-4 border border-gray-300 w-[170px]"
          >
            未
          </td>
        </tr>
        <tr
          class="border-b border-gray-300"
        >
          <td
            class="text-center py-2 px-4 border border-gray-300 w-[70px]"
          >
            <span />
          </td>
          <td
            class="text-center py-2 px-4 border border-gray-300  w-[180px] hover:text-gray-400 underline underline-offset-2"
          >
            <button
              class="cursor-pointer w-full text-center"
            >
              田中一郎
            </button>
          </td>
          <td
            class="text-center py-2 px-4 border border-gray-300 w-[213px]"
          >
            07011112222
          </td>
          <td
            class="text-center py-2 px-4 border border-gray-300 w-[150px]"
          >
            神奈川県
          </td>
          <td
            class="text-center py-2 px-4 border border-gray-300 w-[170px]"
          >
            済
          </td>
        </tr>
      </tbody>
    </table>
    <div
      class="w-full text-center my-7"
    >
      <div
        data-testid="pagination-mock"
      >
        Pagination: limit=10 data-length=3
      </div>
      <button
        class="cursor-pointer  
                
                
                bg-blue-500 rounded-md px-4 py-2 mr-6 mt-4 text-white
                 hover:bg-blue-600  
                "
        type="button"
      >
        閉じる
      </button>
      <button
        class="cursor-pointer bg-blue-500 rounded-md px-4 py-2 mr-6 mt-4 text-white  
                            bg-gray-400 !cursor-not-allowed opacity-50"
        disabled=""
      >
        追加
      </button>
    </div>
  </div>
</DocumentFragment>
`;
